File: java-api-core/src/com/kinvey/java/Metrics.java
package com.kinvey.java;

import java.util.HashMap;

public class Metrics {
	
	private static HashMap<String, Metrics> metricMap;
	
	public static Metrics getMetrics(String className){
		if (metricMap == null){
			metricMap = new HashMap<String, Metrics>();
		}
		if(!metricMap.containsKey(className)){
			metricMap.put(className, new Metrics());
		}
		
		return metricMap.get(className);
	}
	
	
	
	private HashMap<String, Metric> events;
	
	public void start(String event){
		long start = System.currentTimeMillis();
		events.get(event).setStart(start);
	}
	
	public void end(String event){
		long end = System.currentTimeMillis();
		events.get(event).setEnd(end);
			
	}
	
	
	
	private Metrics(){
		events = new HashMap<String, Metrics.Metric>();
	}

	
	
	private static class Metric{
		
		private long start = -1l;
		private long end = -1l;
		private long delta = -1l;
		private boolean calculated = false;
		
		private Metric(){}

		public long getStart() {
			return start;
		}

		public void setStart(long start) {
			this.start = start;
		}

		public long getEnd() {
			return end;
		}

		public void setEnd(long end) {
			this.end = end;
			if (this.start == -1l){
				return;
			}
			this.calculated = true;
			this.delta = this.end - this.start;	
		}

		public long getDelta() {
			return delta;
		}

		public boolean isCalculated() {
			return calculated;
		}		
	}
	
	
	

}

File: java-api-core/src/com/kinvey/java/Metrics.kt
/*
 *  Copyright (c) 2016, Kinvey, Inc. All rights reserved.
 *
 * This software is licensed to you under the Kinvey terms of service located at
 * http://www.kinvey.com/terms-of-use. By downloading, accessing and/or using this
 * software, you hereby accept such terms of service  (and any agreement referenced
 * therein) and agree that you have read, understand and agree to be bound by such
 * terms of service and are of legal age to agree to such terms with Kinvey.
 *
 * This software contains valuable confidential and proprietary information of
 * KINVEY, INC and is subject to applicable licensing agreements.
 * Unauthorized reproduction, transmission or distribution of this file and its
 * contents is a violation of applicable laws.
 *
 */

package com.kinvey.java

import java.util.*

class Metrics {

    private val events: HashMap<String, Metric> = HashMap()

    private var metricMap: HashMap<String, Metrics>? = null

    fun start(event: String) {
        val start = System.currentTimeMillis()
        events[event]?.start = start
    }

    fun end(event: String) {
        val end = System.currentTimeMillis()
        events[event]?.end = end
    }

    fun getMetrics(className: String): Metrics? {
        if (metricMap == null) {
            metricMap = HashMap()
        }
        metricMap?.let {
            if (!it.containsKey(className)) {
                it[className] = Metrics()
            }
            return it[className]
        }
        return null
    }

    private class Metric {
        var start = -1L
        var end = -1L
            set(end) {
                field = end
                if (this.start == -1L) {
                    return
                }
                isCalculated = true
                delta = this.end - this.start
            }
        var delta = -1L
            private set
        var isCalculated = false
            private set
    }
}

